src/Handler/Search.hs

```diff
@@ -21,26 +24,87 @@ import qualified Language.PureScript as P
 
 import qualified XMLArrows
 
+resultsPerPage :: Int
+resultsPerPage = 50
+
+maxPages :: Int
+maxPages = 10
+
 getSearchR :: Handler TypedContent
 getSearchR = do
   mquery <- lookupGetParam "q"
+  page <- lookupCurrentPage
+
   case mquery of
     Nothing -> redirect HomeR
     Just query -> do
-      results <- case tryParseType query of
-        Just ty | not (isSimpleType ty) -> searchForType ty
-        _ -> searchForName (toLower query)
+      (results, hasMore) <- case tryParseType query of
+        Just ty | not (isSimpleType ty) -> searchForType page ty
+        _ -> searchForName page (toLower query)
       selectRep $ do
-        provideRep (htmlOutput query results)
-        provideRep (jsonOutput results)
+        provideRep (htmlOutput query results page hasMore)
+        provideRep (jsonOutput results page hasMore)
   where
-    htmlOutput :: Text -> [SearchResult] -> Handler Html
-    htmlOutput query results = do
+    getPageLinkRenderer :: Handler (Maybe (Int -> Text))
+    getPageLinkRenderer = runMaybeT $ do
+      cr <- MaybeT getCurrentRoute
+      MaybeT $ do
+        urp <- getUrlRenderParams
+        getParams <- Map.fromList . reqGetParams <$> getRequest
+        return $ Just $ \p ->
+          urp cr $ Map.toList
+                 $ Map.insert "page" (pack $ show p) getParams
+
+    mkPaginationW :: Int -> Bool -> Handler (Maybe Widget)
+    mkPaginationW page hasMore = getPageLinkRenderer >>= \mr -> case mr of
+        Nothing -> return Nothing
+        Just mkPageLink ->
+          let allPages = [1..page]
+              mmNextPageLink = if hasMore
+                                  then Just $ if page >= maxPages
```

Comment:
In the past I would have agreed with you, but every single time I've used `Maybe (Maybe a)` like this, I've regretted it. In my opinion it's a far larger cognitive burden to have to remember what the meaning of `Nothing` vs `Just Nothing` is.

However before you address that, I noticed that this is what page 10 now looks like on my machine:
![screenshot from 2017-07-15 13-28-42](https://user-images.githubusercontent.com/1270186/28239460-a62e31d4-6964-11e7-9c4c-789764335141.png)

and I was thinking maybe it would make more sense to have the "further results have been omitted" message after the last search result rather than inside the pagination widget.

Also I think the clickable regions in the "goto page n" links should still be made a little larger. Also I don't think there should be non-clickable gaps in between the links. It might also be nice to have the background change colour slightly on hover so that it's clear what you're clicking on. Finally I think it might be slightly nicer to have the next page link text include a chevron character like "next page &raquo;" and similar for the prev page link?